tsl/src/fdw/deparse.c

```diff
@@ -867,6 +867,60 @@ deparseDistinctClause(StringInfo buf, deparse_expr_cxt *context)
 			return;
 	}
 
+	if (query->hasDistinctOn)
+	{
+		/*
+		 * Pushing down DISTINCT ON is more complex than plain DISTINCT.
+		 * The DISTINCT ON columns must be a prefix of the ORDER BY columns.
+		 * Without this, the DISTINCT ON would return an unpredictable row
+		 * each time. There is a diagnostic for the case where the ORDER BY
+		 * clause doesn't match the DISTINCT ON clause, so in this case we
+		 * would get an error on the data node. There is no diagnostic for
+		 * the case where the ORDER BY is absent, so in this case we would
+		 * get a wrong result.
+		 * The remote ORDER BY clause is created from the pathkeys of the
+		 * corresponding relation. If the DISTINCT ON columns are not a prefix
+		 * of these pathkeys, we cannot push it down.
+		 */
+		int next_pathkey_index = 0;
+		foreach (l, query->distinctClause)
+		{
+			SortGroupClause *sgc = lfirst_node(SortGroupClause, l);
+			TargetEntry *tle = get_sortgroupclause_tle(sgc, query->targetList);
+
+			if (next_pathkey_index >= list_length(pathkeys))
+			{
+				return;
+			}
+
+			PathKey *pk = list_nth(pathkeys, next_pathkey_index);
```

Comment:
instead of keeping an extra index that you increment, can you loop both lists at the same time with `forboth` instead of `foreach`? It seems that would simplify the code and make it more clear that the two lists are iterated in lockstep.

Reply:
Good idea, it's more concise this way.